target_chains/solana/programs/pyth_solana_receiver/tests/program_test/mod.rs

```diff
@@ -40,72 +40,24 @@ pub struct ProgramSimulator {
     /// transaction; otherwise, replayed transactions in different states can return stale
     /// results.
     last_blockhash:        Hash,
-    programdata_id:        Pubkey,
     pub upgrade_authority: Keypair,
     pub genesis_keypair:   Keypair,
 }
 
 impl ProgramSimulator {
     /// Deploys the target chain contract as upgradable
     pub async fn new() -> ProgramSimulator {
-        let mut bpf_data = read_file(
-            std::env::current_dir()
-                .unwrap()
-                .join(Path::new("../../target/deploy/pyth_solana_receiver.so")),
-        );
-
-
-        let mut program_test = ProgramTest::default();
-        let program_key = Pubkey::try_from(ID).unwrap();
-        // This PDA is the actual address in the real world
-        // https://docs.rs/solana-program/1.6.4/solana_program/bpf_loader_upgradeable/index.html
-        let (programdata_key, _) =
-            Pubkey::find_program_address(&[&program_key.to_bytes()], &bpf_loader_upgradeable::id());
+        let mut program_test = ProgramTest::new("pyth_solana_receiver", ID, None);
 
         let upgrade_authority_keypair = Keypair::new();
 
-        let program_deserialized = UpgradeableLoaderState::Program {
-            programdata_address: programdata_key,
-        };
-        let programdata_deserialized = UpgradeableLoaderState::ProgramData {
-            slot:                      1,
-            upgrade_authority_address: Some(upgrade_authority_keypair.pubkey()),
-        };
-
-        // Program contains a pointer to progradata
-        let program_vec = bincode::serialize(&program_deserialized).unwrap();
-        // Programdata contains a header and the binary of the program
-        let mut programdata_vec = bincode::serialize(&programdata_deserialized).unwrap();
-        programdata_vec.append(&mut bpf_data);
-
-        let program_account = Account {
-            lamports:   Rent::default().minimum_balance(program_vec.len()),
-            data:       program_vec,
-            owner:      bpf_loader_upgradeable::ID,
-            executable: true,
-            rent_epoch: Epoch::default(),
-        };
-        let programdata_account = Account {
-            lamports:   Rent::default().minimum_balance(programdata_vec.len()),
-            data:       programdata_vec,
-            owner:      bpf_loader_upgradeable::ID,
-            executable: false,
-            rent_epoch: Epoch::default(),
-        };
-
-        // Add to both accounts to program test, now the program is deploy as upgradable
-        program_test.add_account(program_key, program_account);
-        program_test.add_account(programdata_key, programdata_account);
```

Comment:
(FYI, we've been doing it this way ^ because it lets us test the program upgrade flow in the simulator. Not sure if the replacement code enables that or not)

tbh we should really pull this out to a rust library somewhere and import it because we've copy/pasted this code like 10 times now.

Reply:
i see that makes sense. I did a quick look into this and one alternative to being able to test program upgrade flows would be to use a ts test with solana-local-validator and submit commands like this https://github.com/Squads-Protocol/squads-mpl/blob/main/tests/squads-mpl.ts

Reply:
Please don't use typescript for the tests.  It's way more annoying and difficult to write the tests that way because you can't rely as directly on the rust types and stuff.